llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp

```diff
@@ -0,0 +1,69 @@
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/Config/llvm-config.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/Compiler.h"
+
+using namespace llvm;
+
+namespace {
+
+struct ForWrapper : PassInfoMixin<ForWrapper> {
+  llvm::PreservedAnalyses run(llvm::Function &F, llvm::FunctionAnalysisManager &FA) {
+
+    auto *FM = FunctionType::get(Type::getVoidTy(F.getContext()), false);
+    auto LoopStartFunction = F.getParent()->getOrInsertFunction("loop_start", FM);
+    auto LoopEndFunction = F.getParent()->getOrInsertFunction("loop_end", FM);
+    IRBuilder<> Builder(F.getContext());
+
+    auto &LA = FA.getResult<LoopAnalysis>(F); // Gets the result of loop analysis
+
+    for (auto &L : LA) {
+      auto *Head = L->getHeader();
+      for (auto *const Pred : children<Inverse<BasicBlock *>>(Head)) {
+        if (!L->contains(Pred)) {
+          Builder.SetInsertPoint(Pred->getTerminator());
```

Comment:
Please insert loop_start/loop_end calls only if they are not present already.
And add a test for that case.